dynamo_db_block.py

```diff
@@ -1,7 +1,185 @@
+
+    def __init__(self):
+        super().__init__()
+        self._conn = None
+        self._table_cache = {}
+        self._table_locks = defaultdict(Lock)
+
+    def configure(self, context):
+        super().configure(context)
+        region_name = re.sub('_', '-', self.region.name)
+        self._logger.debug("Connecting to region {}...".format(region_name))
+        self._conn = connect_to_region(
+            region_name,
+            aws_access_key_id=self.creds.access_key,
+            aws_secret_access_key=self.creds.access_secret)
+        self._logger.debug("Connection complete")
+
+    def process_signals(self, signals, input_id='default'):
+        # Split the signals up into table groups for batch writing
+        # batch_groups is a dictionary where the key is a table name, and the
+        # value is a list of signals that should be saved to that table name
+        batch_groups = defaultdict(list)
+        for sig in signals:
+            batch_groups[self.table(sig)].append(sig)
+        self._logger.debug("Processing {} signals in {} batch groups".format(
+            len(signals), len(batch_groups)))
+
+        # Go through each table in batch groups and batch write all of the
+        # signals to the table
+        for table_name, sigs in batch_groups.items():
+            self._logger.debug("Saving {} signals to {}".format(
+                len(sigs), table_name))
+            try:
+                # Lock around each table - in case it is creating still
+                self._logger.debug(
+                    "Waiting for table lock on {}".format(table_name))
+                with self._table_locks[table_name]:
+                    self._logger.debug(
+                        "Table lock acquired for {}".format(table_name))
+                    table = self._get_table(table_name)
+                    self._save_signals_to_table(table, sigs)
+            except:
+                self._logger.exception("Could not batch write to table {}"
+                                       .format(table_name))
+
+    def _get_table(self, table_name):
+        """ Get a DynamoDB table reference based on a table name.
+
+        If we have looked up/created this table before, use the cached
+        reference. If not, check to make sure the table exists. If it does
+        not exist, create it and return that reference.
+
+        Note that if this function does not find the table, it will create it
+        and this creation operation can block for some time (typically ~10s).
+        It will only return the table reference once the table is active and
+        ready to be stored to or read from.
+
+        Returns:
+            table: A boto table reference
+        """
+        if table_name in self._table_cache:
+            return self._table_cache[table_name]
+
+        table = Table(table_name, connection=self._conn)
+
+        try:
+            num_items = table.count()
```

Comment:
Yeah, exactly. We need to try to do something with the table (there is no `table.exists()` type method unfortunately). Count is nice because it gets cached after the first time too so subsequent calls will return much faster